pkg/blockbuilder/blockbuilder.go

```diff
@@ -242,142 +258,147 @@ func cycleEndAtStartup(interval, buffer time.Duration) time.Time {
-	parts := make(map[int32]partitionLag, len(assignmentParts))
-	groupLag := lags[b.cfg.Kafka.ConsumerGroup].Lag
-	for _, part := range assignmentParts {
-		pl, ok := groupLag.Lookup(b.cfg.Kafka.Topic, part)
+		if b.cancelActivePart != nil {
+			b.cancelActivePart(context.Canceled)
+			b.cancelActivePart = nil
+		}
+
+		b.parts.reset()
+	}()
+
+	for ctx.Err() == nil {
+		b.mu.Lock()
+		ctx, cancel := context.WithCancelCause(ctx)
+		b.cancelActivePart = cancel
+
+		part := b.parts.next()
+		b.mu.Unlock()
+
+		if part < 0 {
+			return nil
+		}
+
+		level.Debug(b.logger).Log("msg", "next consume cycle", "part", part)
+
+		lags, err := kadmClient.Lag(ctx, b.cfg.Kafka.ConsumerGroup)
+		if err != nil {
+			return fmt.Errorf("get consumer group lag: %w", err)
+		} else if err := lags.Error(); err != nil {
+			return fmt.Errorf("get consumer group lag: %w", err)
+		}
+
+		groupLag := lags[b.cfg.Kafka.ConsumerGroup].Lag
+		lag, ok := groupLag.Lookup(b.cfg.Kafka.Topic, part)
 		if !ok {
 			continue
 		}
-		if pl.Lag <= 0 {
-			if pl.Err != nil {
-				level.Error(b.logger).Log("msg", "failed to get partition lag", "err", pl.Err, "part", part)
+
+		if lag.Lag <= 0 {
+			if err := lag.Err; err != nil {
+				level.Error(b.logger).Log("msg", "failed to get partition lag", "err", err, "part", part)
 			} else {
 				level.Info(b.logger).Log(
 					"msg", "nothing to consume in partition",
-					"err", pl.Err,
 					"part", part,
-					"offset", pl.Commit.At,
-					"end_offset", pl.End.Offset,
-					"lag", pl.Lag,
+					"offset", lag.Commit.At,
+					"end_offset", lag.End.Offset,
+					"lag", lag.Lag,
 				)
 			}
 			continue
 		}
 
+		b.metrics.consumerLag.WithLabelValues(lag.Topic, fmt.Sprintf("%d", lag.Partition)).Set(float64(lag.Lag))
+
 		// We look at the commit offset timestamp to determine how far behind we are lagging
 		// relative to the cycleEnd. We will consume the partition in parts accordingly.
-		offset := pl.Commit
-		level.Debug(b.logger).Log("part", offset.Partition, "offset", offset.At, "meta", offset.Metadata)
+		offset := lag.Commit
 		commitRecTs, seenTillTs, lastBlockEnd, err := unmarshallCommitMeta(offset.Metadata)
 		if err != nil {
 			// If there is an error in unmarshalling the metadata, treat it as if
 			// we have no commit. There is no reason to stop the cycle for this.
 			level.Warn(b.logger).Log("msg", "error unmarshalling commit metadata", "err", err, "part", part, "offset", offset.At, "metadata", offset.Metadata)
 		}
 
-		b.metrics.consumerLag.WithLabelValues(pl.Topic, fmt.Sprintf("%d", pl.Partition)).Set(float64(pl.Lag))
-
-		parts[part] = partitionLag{
+		pl := partitionLag{
 			Partition: part,
-			Lag:       pl.Lag,
+			Lag:       lag.Lag,
 
 			CommitRecTs:  commitRecTs,
 			SeenTillTs:   seenTillTs,
 			LastBlockEnd: lastBlockEnd,
 		}
+		if err := b.nextConsumeCycle(ctx, pl, cycleEnd); err != nil {
+			level.Error(b.logger).Log("msg", "failed to consume partition", "err", err, "part", part)
+		}
 	}
+	return nil
```

Comment:
I think we should return `ctx.Err()` here. And if it was the partition lost error, then the `running()` should retry `NextConsumeCycle()` with the new partitions. Else there might be some partitions that will lag for a cycle. (can we test this somehow?)

Reply:
Caution: at this point, `ctx.Err()` only happens when that service is stopping. Note that the `partition lost` ONLY happens inside the loop above this line.

I think we need a better name for the `ctx` on L285 so it was less confusing.

Reply:
Yup. Explicitly different names would be good. Even if it is working correctly now, it will be easy to introduce bugs without noticing.